### PR TITLE
Use GHC_ENVIRONMENT to resolve module name ambiguity in doctests

### PR DESCRIPTION
**Pull request by qrilka**
Note: Documentation fixes for https://docs.haskellstack.org/en/stable/ should target the "stable" branch, not master.

Please include the following checklist in your PR:

* [x] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [ ] The documentation has been updated, if necessary.

Tested manually on an intermediate Stackage snapshot (`lens` vs `lenz` ambiguity in `language-nix` tests)
This overrides #4533 and makes  https://github.com/sol/doctest/pull/218 unnecessary

### CODE REVIEW

File: src/Stack/Build/Execute.hs [hunk 1]
```diff
@@ -1809,22 +1810,39 @@ singleTest topts testsToRun ac ee task installedMap = do
...
-                let usePackageIds pc = modifyEnvVars pc $ \envVars ->
-                      Map.insert "HASKELL_PACKAGE_IDS" (T.unwords packageIds) envVars
-                menv <- liftIO $ usePackageIds =<< configProcessContextSettings config EnvSettings
+                let setEnv f pc = modifyEnvVars pc $ \envVars ->
+                      Map.insert "GHC_ENVIRONMENT" (T.pack f) envVars
+                    fp = toFilePath $ eeTempDir ee </> $(mkRelFile "test-ghc-env")
+                    snapDBPath = toFilePathNoTrailingSep (bcoSnapDB $ eeBaseConfigOpts ee)
+                    localDBPath = toFilePathNoTrailingSep (bcoLocalDB $ eeBaseConfigOpts ee)
+                    ghcEnv = mconcat $
+                        "clear-package-db\n" :
```
Comment (snoyberg) [inline src/Stack/Build/Execute.hs]: It's slightly more efficient to bypass the `mconcat` and do `"foo" <> "bar" <> ... <> foldMap f rest`, but that's not terribly important here.

Comment (qrilka) [reply src/Stack/Build/Execute.hs]: I'll fix it anyway

**Comment by qrilka** [comment]
@snoyberg I think `HASKELL_PACKAGE_SANDBOX`(`ES`) is also aimed at `doctest` and probably could be removed. Also with that change the code in https://github.com/sol/doctest/blob/master/src/PackageDBs.hs could be needed only for backward compatibilty. But let's do it in a separate PR (or maybe good cleanup ticket for a newcomer?).

File: src/Stack/Build/Execute.hs (hunk 1 above)
Comment (qrilka) [reply src/Stack/Build/Execute.hs]: Done in 99059212

**Pull request merged**